Improve the pygments lexer to cover GDScript 2 better

Qubus0 opened the pull request:
we've [recently updated the gdscript lexer](https://github.com/GodotModding/pygments-gdscript) for our own documentation, this is a port of those changes to fit the godot docs

please check as many pages as you can for errors i might have missed. 

also let me know if the colors are fine like this. i think some of the existing colors may be from an older theme and not quite up to date, but i didn't touch them for now.

main changes are functions having proper blue color now, global functions being purple, all classes getting the correct greens (including builtins and types having the different green), and proper colors for $String, string names and node paths

Qubus0 commented:
Side by side

<img width="3260" alt="image" src="https://github.com/user-attachments/assets/b0edc4b2-1b5e-4d12-9ffd-7da428576d6c" />

<img width="3296" alt="LexerUpdate2" src="https://github.com/user-attachments/assets/22fa93bc-3413-4749-acf8-0ecc9259fe44" />

<img width="3264" alt="LexerUpdate3" src="https://github.com/user-attachments/assets/d6233b13-8492-4d16-a661-2c63057d619d" />

Calinou commented on _extensions/gdscript.py:
```diff
@@ -9,45 +9,37 @@
     name = "GDScript"
-    aliases = ["gdscript", "gd"]
+    url = "https://www.godotengine.org"
+    aliases = ["gdscript", "gd", "gdscript2", "gd2"]
```
I would remove the aliases we don't use:

```suggestion
    aliases = ["gdscript", "gd"]
```

This lexer should still do a good enough job with Godot 3.x GDScript code anyway, so the distinction isn't really relevant.

Qubus0 replied:
ah yep, we're running the full own lexer here. in our case we had a conflict with the builtin pygments lexer forcing us to add an alias. will remove

Qubus0 commented on _extensions/gdscript.py:
```diff
@@ -440,50 +1406,172 @@ def innerstring_rules(ttype):
             (
-                r"(-)?((\d|(?<=\d)_)+\.(\d|(?<=\d)_)*|(\d|(?<=\d)_)*\.(\d|(?<=\d)_)+)([eE][+-]?(\d|(?<=\d)_)+)?j?",
-                Number.Float,
+                r"!=|==|<<|>>|&&|\+=|-=|\*=|/=|%=|&=|\|=|\|\||:=|[-~+/*%=<>&^.!|$]",
```
i guess removing it won't make a big difference since `=` is covered as single operator and `:` is covered by punctuation

Qubus0 commented:
Name and Punctuation didn't have css classes yet. adding them makes the () and = bluish, pretty hard to notice, and fixes code in warning callouts being yellow

<img width="3269" height="1824" alt="image" src="https://github.com/user-attachments/assets/0d5e83f7-0063-428c-9c1f-8140576e92e7" />

Calinou commented on _extensions/gdscript.py:
```diff
@@ -440,50 +1406,172 @@ def innerstring_rules(ttype):
             ),
         ],
-        "numbers": [
+        "operator": [
```
I would use singular everywhere. Most highlighters seem to do this in their local definitions.

Calinou reviewed (commented):
Looks great now!

Dark | Light
-|-
<img width="825" height="782" alt="Screenshot_20250829_225651" src="https://github.com/user-attachments/assets/272ad8ad-c105-403f-91db-d490548157d5" /> | <img width="825" height="782" alt="Screenshot_20250829_225706" src="https://github.com/user-attachments/assets/f27f065b-44d4-4f71-93da-5ed96d960472" />

I can confirm the code color in admonitions is fixed now.

Qubus0 commented:
singulars done, squashed everything, should be ready - let me know if anything else stands out to you

Calinou commented on _extensions/gdscript.py:
```diff
@@ -148,27 +81,27 @@ def innerstring_rules(ttype):
                         "self",
                         "super",
-                        "yield", # Reserved for potential future use.
+                        "null",  # type, but colored like a keyword
```
This matches the GDScript editor behavior, so I d say it makes sense.

Qubus0 commented:
Updated preview
<details>

| dark | light |
|---|---|
|<img alt="image" src="https://github.com/user-attachments/assets/b6d8d87a-2d59-4ab2-9d7f-abee7074b1bd" />|<img  alt="image" src="https://github.com/user-attachments/assets/c719d57b-0cc0-413a-81ad-20c6de8cb6d9" />|
|<img alt="image" src="https://github.com/user-attachments/assets/46f9775b-d81d-4002-ba95-288ec3ca9620" />|<img alt="image" src="https://github.com/user-attachments/assets/198023d0-81f3-4276-9f47-562df03882e8" />|
|<img alt="image" src="https://github.com/user-attachments/assets/4ef8b8a2-a39d-4247-84d0-be4e45644b88" />|<img alt="image" src="https://github.com/user-attachments/assets/ae4376eb-0cf3-48be-8bb3-ed9a3cccbc33" />|
|<img alt="image" src="https://github.com/user-attachments/assets/d173a3fe-4dbc-4b01-a196-743993794760" />|<img alt="image" src="https://github.com/user-attachments/assets/4748a6a3-ea6c-4c57-a922-6858d6658d5a" />|

</details>

Qubus0 commented:
fixed member vars color and `...args` 
|dark|light|
|---|---|
|<img width="354" height="126" alt="image" src="https://github.com/user-attachments/assets/307346c6-efd7-4fa1-94f6-9ca2b159f67b" />|<img width="311" height="127" alt="image" src="https://github.com/user-attachments/assets/040338ec-5635-4fbb-8fab-8607dd1f6c54" />|

code not changed, just added for simpler screenshot

mhilbrunner commented on _extensions/gdscript.py:
```diff
@@ -391,8 +333,1032 @@ def innerstring_rules(ttype):
             ),
-        ],
-        "decorators": [
+            # copied from https://docs.godotengine.org/en/stable/classes/index.html
+            (
+                words(
+                    (
+                        # Nodes
+                        "Node",
+                        "AcceptDialog",
+                        "AnimatableBody2D",
+                        "AnimatableBody3D",
```
I have concerns about this. It feels like this will grow out of date quickly and would need constant updates. If there is any way at all to automate this part, I'd appreciate it.

Qubus0 replied:
While I agree automation would be nice (and I'll have a look into it), realistically I don't think this going out of date is much of an issue because 

1. The most important classes have had the same name since 4.x, they aren't renamed particularly often. 
2. New classes are rarely added to the docs right away
3. Classes are rarely removed between major updates too
4. If there are any outdated classes, the hue will only be off slightly
5. And considering all of that together, the benefit of covering most of the important classes far outweighs  risking missing a few less important ones

mhilbrunner replied:
Those are some fair points. :) Maybe we can make grabbing the list from that URL during the build a future todo then, and see how it goes.
